pkg/scheduler/queue/user_queues.go

```diff
@@ -28,298 +27,350 @@ type querier struct {
+	//  - sharding is disabled (max-queriers-per-tenant=0)
+	//  - or if max-queriers-per-tenant >= the number of queriers
+	//
+	// Tenant -> Queriers is the core relationship randomized from the shuffle shard seed.
+	// The shuffle shard seed is itself consistently hashed from the tenant ID.
+	// However, the most common operation is the querier asking for its next request,
+	// which requires a relatively efficient lookup or check of Querier -> Tenant.
+	//
+	// Reshuffling is done when:
+	//  - a querier connection is added or removed
+	//  - it is detected during request enqueueing that a tenant's queriers
+	//    were calculated from an outdated max-queriers-per-tenant value
+
+	queriersByID map[string]*querierConn
+	// Sorted list of querier ids, used when shuffle sharding queriers for tenant
+	querierIDsSorted []string
 
 	// How long to wait before removing a querier which has got disconnected
 	// but hasn't notified about a graceful shutdown.
-	forgetDelay time.Duration
+	querierForgetDelay time.Duration
+
+	// List of all tenants with queues, used for iteration when searching for next queue to handle.
+	// Tenants removed from the middle are replaced with "". To avoid skipping users during iteration,
+	// we only shrink this list when there are ""'s at the end of it.
+	tenantIDOrder []string
+	tenantsByID   map[string]*queueTenant
+
+	// Tenant assigned querier ID set as determined by shuffle sharding.
+	// If tenant querier ID set is not nil, only those queriers can handle the tenant's requests,
+	// Tenant querier ID is set to nil if sharding is off or available queriers <= tenant's maxQueriers.
+	tenantQuerierIDs map[string]map[string]struct{}
+}
+
+type queueTenant struct {
+	// seed for shuffle sharding of queriers; computed from userID only,
+	// and is therefore consistent between different frontends.
+	shuffleShardSeed int64
 
-	// Tracks queriers registered to the queue.
-	queriers map[string]*querier
+	// points up to user order to enable efficient removal
+	orderIndex int
 
-	// Sorted list of querier names, used when creating per-user shard.
-	sortedQueriers []string
+	maxQueriers int
 }
 
-type userQueue struct {
-	requests *list.List
+// This struct holds user queues for pending requests. It also keeps track of connected queriers,
+// and mapping between users and queriers.
+type queues struct {
+	tenantQueues map[string]*userQueue
 
-	// If not nil, only these queriers can handle user requests. If nil, all queriers can.
-	// We set this to nil if number of available queriers <= maxQueriers.
-	queriers    map[string]struct{}
-	maxQueriers int
+	tenantQuerierState tenantQuerierState
 
-	// Seed for shuffle sharding of queriers. This seed is based on userID only and is therefore consistent
-	// between different frontends.
-	seed int64
+	maxUserQueueSize int
+}
 
-	// Points back to 'users' field in queues. Enables quick cleanup.
-	index int
+type userQueue struct {
+	requests *list.List
 }
```

Comment:
this could be flattened so it's no longer a separate type at all

Reply:
When you say flattened, do you mean making this something like `type userQueue *list.List`? If so, that makes sense to me.

Reply:
I think I may keep it unflattened for now, for when we move to the newer queue implementation.

That way I can wrap both userQueue and the new queue a a basic queue interface so they can be swapped 1:1.